src/ssl.jl

```diff
@@ -643,18 +674,18 @@ function Base.close(ssl::SSLStream, shutdown::Bool=true)
     # if we've already finalized, no further action needed
     ssl.ssl.ssl == C_NULL && return
     # close operations
-    Base.@lock ssl.lock begin
+    # close underlying io
+    try
+        Base.close(ssl.io)
+    catch e
+        e isa Base.IOError || rethrow()
+    end
+    Base.@lock ssl.closelock begin
```

Comment:
I'm not sure why we need a lock here -- IIUC, everything behind the lock should happen only once per the lifetime of the SSLStream, in which case we should be able to avoid a lock an just use atomics a la:
```julia
Base.close(ssl::SSLStream, shutdown::Bool=true)
    closed = @atomic ssl.closed
    closed && return
    if (@atomicswap ssl.closed false true).success
        ssl.ssl.ssl == C_NULL && return nothing
        # close operations
        # close underlying io
        try
            Base.close(ssl.io)
        catch e
            e isa Base.IOError || rethrow()
        end
        # Ignore the disconnect result.
        shutdown && ssl_disconnect(ssl.ssl)
        free(ssl.ssl)
    end
    return nothing
end
```

I'm also not sure when we'd want to close the SSLStream but not call `ssl_disconnect`

Reply:
Yeah, I think you're right. Thanks for checking on this.

We need to avoid calling `ssl_disconnect` in the case we get an `ERROR_SSL_SYSCALL` or a few other return codes as per openssl documentation. i.e. if something "really bad" seems to happen during a read/write operation, then they say to _free_ the SSL, but don't call disconnect.

Reply:
Thanks! Would be great to mention this motivation in a docstring / comment

Reply:
Ok, pushed a commit to fixup